### PR TITLE
Unique ptr

### PR DESCRIPTION
**Pull request by NeoBauer**

### CODE REVIEW

File: homework/unique_ptr/uniqueptr.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#pragma once
+#include <stdexcept>
+#include <stdio.h>
+#include <iostream>
+
+template <class T> class unique_ptr {
```
Comment (jakjan95) [inline homework/unique_ptr/uniqueptr.hpp:6]: instead of class use typename in template

File: homework/unique_ptr/uniqueptr.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  T *get() const;
+  T release();
+  void reset();
+  T &operator*();
+
+private:
+  T *rawPtr_;
+};
+
```
Comment (jakjan95) [inline homework/unique_ptr/uniqueptr.hpp:23]: I would rather use that kind of naming convention: T* something, T& something :)

File: homework/unique_ptr/uniqueptr.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+template <class T> unique_ptr<T>::unique_ptr(T *rawPtr) : rawPtr_(rawPtr) {
+  puts("Yes, we created uniqueptr! :)");
```
Comment (jakjan95) [inline homework/unique_ptr/uniqueptr.hpp:25]: puts() is bad habit in C so i would not use it in C++

File: homework/unique_ptr/uniqueptr.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    : rawPtr_(other.rawPtr_)
+
+{
+  other.rawPtr_ = nullptr;
+}
+
+template <class T>
+unique_ptr<T>::~unique_ptr() // destructor
+{
+  std::cout << "Destructor called where raw_ptr adress is: " << this->rawPtr_ << '\n';
```
Comment (jakjan95) [inline homework/unique_ptr/uniqueptr.hpp:39]: cout can throw exception so it is not good idea to put it here

File: homework/unique_ptr/uniqueptr.hpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+template <class T> T unique_ptr<T>::release() {
+  auto result = rawPtr_;
+  rawPtr_ = nullptr;
+  return result;
+}
+
+template <class T> void unique_ptr<T>::reset() { rawPtr_ = nullptr; }
+template <class T> T &unique_ptr<T>::operator*() { return *rawPtr_; }
```
Comment (jakjan95) [inline homework/unique_ptr/uniqueptr.hpp:57]: what about nullptr dereference?

**Review by jakjan95 — CHANGES_REQUESTED** [review]
Code need little improvement and it will be fine :)

File: homework/unique_ptr/uniqueptr.hpp (hunk 2 above)
Comment (NeoBauer) [reply homework/unique_ptr/uniqueptr.hpp:23]: > I would rather use that kind of naming convention: T* something, T& something :)

May I ask you for an example?

File: homework/unique_ptr/uniqueptr.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+template <typename T> T unique_ptr<T>::release() {
+  T* result = rawPtr_;
+  rawPtr_ = nullptr;
+  return result;
+}
+
+template <typename T> void unique_ptr<T>::reset(T *NewRawPtr) noexcept {
+  delete rawPtr_;
```
Comment (MaciejSalamonski) [inline homework/unique_ptr/uniqueptr.hpp:56]: Could you explain that?

**Review by MaciejSalamonski — COMMENTED** [review]
Hi i have one question :P

File: homework/unique_ptr/uniqueptr.hpp (hunk 6 above)
Comment (NeoBauer) [reply homework/unique_ptr/uniqueptr.hpp:56]: Where do you see the problem?

**Review by ziobron — CHANGES_REQUESTED** [review]
There are no unit tests. Without them you cannot prove that your code is correct. Please also add a job that run tests under valgrind or compile them with asan.

File: .github/workflows/UniquePtr.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        make
+    - name: Run tests
+      working-directory: homework/unique_ptr/build
+      run: ctest -V 
+    - name: Run Valgrind
+      working-directory: homework/unique_ptr/build
+      run: |
+          sudo apt-get update -y    
+          sudo apt-get install -y valgrind
+          valgrind --leak-check=yes ./UniquePointerExe
```
Comment (ziobron) [inline .github/workflows/UniquePtr.yml:28]: run UTs under valgrind. Main executable is optional.

File: homework/unique_ptr/tests/tests.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    WHEN("Reseting unique_ptr with new object and value") {
+      firstUnique.reset(new TestClass(input2));
+      THEN("new value in unique_ptr, and earlier called get() is empty") {
+        REQUIRE(firstUnique.get()->data_ == input2);
+      }
+    }
+
+    WHEN("Reseted uniqueptr with nullptr") {
+      firstUnique.reset(nullptr);
```
Comment (ziobron) [inline homework/unique_ptr/tests/tests.cpp:25]: use empty reset call - `firstUnique.reset();`

File: homework/unique_ptr/tests/tests.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    WHEN("Moving into another null_ptr ") {
+      unique_ptr<TestClass> secondUnique = std::move(firstUnique);
+      THEN("Check if there are new parameters") {
+        REQUIRE(firstUnique.get() == nullptr);
+        REQUIRE(secondUnique.get()->data_ == input1);
+      }
+    }
+  }
+}
```
Comment (ziobron) [inline homework/unique_ptr/tests/tests.cpp:44]: tests for release() and some dereference cases are missing

File: homework/unique_ptr/uniqueptr.hpp [hunk 10]
```diff
@@ -12,7 +12,7 @@ template <typename T> class unique_ptr {
   unique_ptr &operator=(const unique_ptr &) = delete; // copy assigment operator
```
Comment (ziobron) [inline homework/unique_ptr/uniqueptr.hpp:12]: what about move assignment operator? It's missing.

File: homework/unique_ptr/uniqueptr.hpp [hunk 11]
```diff
@@ -12,7 +12,7 @@ template <typename T> class unique_ptr {
...
   ~unique_ptr();
 
-  T *get() const;
+  T* get() const;
   T release();
```
Comment (ziobron) [inline homework/unique_ptr/uniqueptr.hpp:16]: release should return T*

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/unique_ptr/main.cpp [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+//PtrShop.push_back(firstUnique); // will not work, because copy constructor is deleted!
+if(firstUnique.get() == nullptr){
+std::cout << "We removed raw_ptr from unique_ptr!" <<'\n';
+}
+if(PtrShop.at(0).get() != nullptr){
+int dataFromShop = PtrShop.at(0).get()->data_;
+std::cout << "Now data from our unique ptr is in vector: " << dataFromShop << "!\n";
+}
+std::cout << "raw_ptr adress stored in vector's unique_ptr after std::move(): " << firstUnique.get() << '\n';
+// std::cout << "Data stored in firstUnique: " << (*firstUnique).data_ << '\n'; // dereferencing nullptr!
```
Comment (WBigaj) [inline homework/unique_ptr/main.cpp:21]: Is this comment needed?

File: homework/unique_ptr/main.cpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+#include <vector>
+#include <iostream>
+
+
+
+int main(){
+unique_ptr<TestClass> firstUnique( new TestClass(2.0));
+std::vector<unique_ptr<TestClass>> PtrShop;
+PtrShop.push_back(std::move(firstUnique));
+//PtrShop.push_back(firstUnique); // will not work, because copy constructor is deleted!
```
Comment (WBigaj) [inline homework/unique_ptr/main.cpp:12]: Is this comment needed?

File: homework/unique_ptr/uniqueptr.hpp [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+template <typename T>
+unique_ptr<T>::~unique_ptr() // destructor
+{
+  delete rawPtr_;
+}
+
+template <typename T>
+T* unique_ptr<T>::get() const { /* if(rawPtr_ == nullptr)
+                                 throw std::range_error("Trying to access empty
+                                 unique_ptr using .get()"); else  */
```
Comment (WBigaj) [inline homework/unique_ptr/uniqueptr.hpp:45]: Is this comment needed?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and there is some unremoved comments.

File: homework/unique_ptr/tests/tests.cpp [hunk 15]
```diff
@@ -21,24 +21,67 @@ SCENARIO("Working on home implemented unique_ptr") {
...
       THEN("unique_ptr is nullptr") { REQUIRE(firstUnique.get() == nullptr); }
     }
-    WHEN("Moving into another container") {
+    WHEN("Reseted unique_ptr with reset()") {
+      firstUnique.reset();
+      THEN("unique_ptr is nullptr") { REQUIRE(firstUnique.get() == nullptr); }
+    }
+    WHEN("Releasing owned pointer without assigning return") {
+      firstUnique.release();
+      THEN("unique_ptr is nullptr") { REQUIRE(firstUnique.get() == nullptr); }
```
Comment (ziobron) [inline homework/unique_ptr/tests/tests.cpp:34]: And we have a memory leak

**Review by ziobron — CHANGES_REQUESTED** [review]
There is one memory leak: https://github.com/NeoBauer/memory-management/runs/1151048352?check_suite_focus=true
Check valgrind output

File: homework/unique_ptr/tests/tests.cpp [hunk 16]
```diff
@@ -29,8 +29,8 @@ SCENARIO("Working on home implemented unique_ptr") {
       firstUnique.reset();
       THEN("unique_ptr is nullptr") { REQUIRE(firstUnique.get() == nullptr); }
     }
-    WHEN("Releasing owned pointer without assigning return") {
-      firstUnique.release();
+    WHEN("Releasing owned pointer with assigning return to another unique_ptr") {
+      unique_ptr<TestClass> containerForReleased = firstUnique.release();
```
Comment (ziobron) [inline homework/unique_ptr/tests/tests.cpp:33]: or just call delete explicitly on returned raw pointer

**Review by ziobron — APPROVED** [review]
👍

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**